### PR TITLE
Fix #12433 Refactor STAFF_TYPE to use strong enum

### LINKED ISSUE #12433: Refactor STAFF_TYPE to use strong enum

### PR DESCRIPTION
**Pull request by ryan-bello**
Changed to a strongly typed enum called StaffType.

To prevent conflict and confusion with peep->StaffType, I renamed it to peep->AssignedStaffType

Tested and works fine.

### CODE REVIEW

**Comment by ryan-bello** [comment]
Sorry about that, I'll fix it and resubmit

**Comment by duncanspumpkin** [comment]
You don't need to close a pr just cause it's not perfect. Switch it to a draft pr until it's ready for review. You end up spamming my email opening and closing PRs and if we have started reviewing it and you open a new pr our review information can be lost. As for clang format just follow the ci's changes or use clang format on your platform. Most IDEs have it built in or a plugin for it built in. Visual studio you can activate it on an open file with ctrl + k followed by ctrl + d.

File: src/openrct2/actions/StaffHireNewAction.hpp [hunk 1]
```diff
@@ -54,13 +54,13 @@ DEFINE_GAME_ACTION(StaffHireNewAction, GAME_COMMAND_HIRE_NEW_STAFF_MEMBER, Staff
 {
 private:
     bool _autoPosition = false;
-    uint8_t _staffType = STAFF_TYPE::STAFF_TYPE_COUNT;
+    StaffType _staffType = StaffType::Count;
```
Comment (duncanspumpkin) [inline src/openrct2/actions/StaffHireNewAction.hpp:57]: See comments on discord. Keep this a uint8_t but keep the constructor with StaffType for the parameter. Cast it in the body of the constructor.

File: src/openrct2/actions/StaffHireNewAction.hpp [hunk 2]
```diff
@@ -77,7 +77,10 @@ DEFINE_GAME_ACTION(StaffHireNewAction, GAME_COMMAND_HIRE_NEW_STAFF_MEMBER, Staff
     {
         GameAction::Serialise(stream);
 
-        stream << DS_TAG(_autoPosition) << DS_TAG(_staffType) << DS_TAG(_entertainerType) << DS_TAG(_staffOrders);
+        uint8_t casted_staffType = static_cast<uint8_t>(_staffType); //there's an error when directly casting in the DS_TAG() function for some reason
```
Comment (duncanspumpkin) [inline src/openrct2/actions/StaffHireNewAction.hpp:80]: You can't do this as you will break multiplayer. Staff type will no longer get assigned.

Comment (ryan-bello) [reply src/openrct2/actions/StaffHireNewAction.hpp:80]: Got it, thank you! I'll fix it soon. And sorry about opening and closing, I'm pretty new to this. I'll reopen and change to a draft.

**Comment by duncanspumpkin** [comment]
Why did you change the member variable to assigned staff type? I assume it was to prevent a name collision?

**Comment by ryan-bello** [comment]
You mean the AssignedStaffType in the Peep struct, right? Yeah, because the strong enum I created was also named StaffType. I could have named it something like StaffTypeEnum, of course, but I thought it would be nice for it to match the PeepType enum, and the AssignedPeepType member variable. Hope that's fine.

**Comment by tupaschoal** [comment]
Note: You said you're pretty new to this, so do read about `git rebase` and squashing commits on rebase, just to make sure you don't throw away anything

**Comment by ryan-bello** [comment]
Yeah, its just those two changes really, everything else was just fixing errors. I'll read up on git rebase and squashing, and I'll push it again after fixing it.

**Comment by ryan-bello** [comment]
Got it down to two commits, both compile and run.

File: src/openrct2-ui/windows/Staff.cpp [hunk 3]
```diff
@@ -887,9 +887,11 @@ void window_staff_options_invalidate(rct_window* w)
             w->pressed_widgets &= ~((1 << WIDX_CHECKBOX_1) | (1 << WIDX_CHECKBOX_2));
             w->pressed_widgets |= peep->StaffOrders << WIDX_CHECKBOX_1;
             break;
-        case STAFF_TYPE_SECURITY:
+        case StaffType::Security:
             // Security guards don't have an options screen.
             break;
+        case StaffType::Count: // All enumeration values must be added to prevent compilation error
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Staff.cpp:893]:
```suggestion
        case StaffType::Count:
```
No need for this comment

File: src/openrct2-ui/windows/Staff.cpp [hunk 4]
```diff
@@ -1174,14 +1176,20 @@ void window_staff_stats_paint(rct_window* w, rct_drawpixelinfo* dpi)
...
             screenCoords.y += LIST_ROW_HEIGHT;
             gfx_draw_string_left(
                 dpi, STR_STAFF_STAT_RIDES_FIXED, static_cast<void*>(&peep->StaffRidesFixed), COLOUR_BLACK, screenCoords);
             break;
+        case StaffType::Security: // All enumeration values must be added to prevent compilation error
+            break;
+        case StaffType::Entertainer:
+            break;
+        case StaffType::Count:
+            break;
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Staff.cpp:1192]:
```suggestion
        case StaffType::Security:
        case StaffType::Entertainer:
        case StaffType::Count:
            break;
```

File: src/openrct2-ui/windows/StaffList.cpp [hunk 5]
```diff
@@ -726,7 +728,7 @@ void window_staff_list_scrollpaint(rct_window* w, rct_drawpixelinfo* dpi, int32_
             }
 
             auto staffOrderIcon_x = nameColumnSize + 20;
-            if (peep->AssignedStaffType != 3)
+            if (static_cast<uint8_t>(peep->AssignedStaffType) != 3)
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/StaffList.cpp:731]:
```suggestion
            if (peep->AssignedStaffType != StaffType::Entertainer)
```

File: src/openrct2/scripting/ScEntity.hpp [hunk 6]
```diff
@@ -1087,14 +1087,16 @@ namespace OpenRCT2::Scripting
...
-                    case STAFF_TYPE_MECHANIC:
+                    case StaffType::Mechanic:
                         return "mechanic";
-                    case STAFF_TYPE_SECURITY:
+                    case StaffType::Security:
                         return "security";
-                    case STAFF_TYPE_ENTERTAINER:
+                    case StaffType::Entertainer:
                         return "entertainer";
+                    case StaffType::Count: // This shouldn't come up, but it's necessary to prevent compilation errors
```
Comment (tupaschoal) [inline src/openrct2/scripting/ScEntity.hpp:1098]:
```suggestion
                    case StaffType::Count:
```

**Comment by ryan-bello** [comment]
Got it, will do! I'll rebase, add myself and clean up those comments. Thanks you guys for your patience and help!

File: src/openrct2/actions/StaffSetColourAction.hpp [hunk 7]
```diff
@@ -44,12 +44,14 @@ DEFINE_GAME_ACTION(StaffSetColourAction, GAME_COMMAND_SET_STAFF_COLOUR, GameActi
     void Serialise(DataSerialiser & stream) override
     {
         GameAction::Serialise(stream);
+
```
Comment (tupaschoal) [inline src/openrct2/actions/StaffSetColourAction.hpp:47]: Can you revert this change? It's not related to the PR (I think it's a leftover from a previous one you did)
```suggestion
```

File: src/openrct2/actions/StaffSetColourAction.hpp [hunk 8]
```diff
@@ -44,12 +44,14 @@ DEFINE_GAME_ACTION(StaffSetColourAction, GAME_COMMAND_SET_STAFF_COLOUR, GameActi
...
         stream << DS_TAG(_staffType) << DS_TAG(_colour);
     }
 
     GameActionResult::Ptr Query() const override
     {
-        if (_staffType != STAFF_TYPE_HANDYMAN && _staffType != STAFF_TYPE_MECHANIC && _staffType != STAFF_TYPE_SECURITY)
+        if (_staffType != static_cast<uint8_t>(StaffType::Handyman) && _staffType != static_cast<uint8_t>(StaffType::Mechanic)
+            && _staffType != static_cast<uint8_t>(StaffType::Security))
```
Comment (tupaschoal) [inline src/openrct2/actions/StaffSetColourAction.hpp:54]: Maybe this looks more readable as:
```suggestion
        auto staffType = static_cast<StaffType>(_staffType);
        if (staffType != StaffType::Handyman && staffType != StaffType::Mechanic && staffType != StaffType::Security)
```

**Comment by ryan-bello** [comment]
adjustments added

File: src/openrct2-ui/windows/Staff.cpp [hunk 9]
```diff
@@ -886,9 +886,11 @@ void window_staff_options_invalidate(rct_window* w)
             w->pressed_widgets &= ~((1 << WIDX_CHECKBOX_1) | (1 << WIDX_CHECKBOX_2));
             w->pressed_widgets |= peep->StaffOrders << WIDX_CHECKBOX_1;
             break;
-        case STAFF_TYPE_SECURITY:
+        case StaffType::Security:
             // Security guards don't have an options screen.
             break;
+        case StaffType::Count:
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Staff.cpp:892]: Just add a default. Staff Type Count should never be an option in a switch

File: src/openrct2-ui/windows/Staff.cpp [hunk 10]
```diff
@@ -1173,14 +1175,18 @@ void window_staff_stats_paint(rct_window* w, rct_drawpixelinfo* dpi)
...
-        case STAFF_TYPE_MECHANIC:
+        case StaffType::Mechanic:
             gfx_draw_string_left(
                 dpi, STR_STAFF_STAT_RIDES_INSPECTED, static_cast<void*>(&peep->StaffRidesInspected), COLOUR_BLACK,
                 screenCoords);
             screenCoords.y += LIST_ROW_HEIGHT;
             gfx_draw_string_left(
                 dpi, STR_STAFF_STAT_RIDES_FIXED, static_cast<void*>(&peep->StaffRidesFixed), COLOUR_BLACK, screenCoords);
             break;
+        case StaffType::Security:
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/Staff.cpp:1186]: Add a default

File: src/openrct2/peep/Peep.h [hunk 11]
```diff
@@ -60,6 +60,16 @@ enum class PeepType : uint8_t
     Invalid = 0xFF
 };
 
+enum class StaffType : uint8_t
```
Comment (duncanspumpkin) [inline src/openrct2/peep/Peep.h:63]: Move back to staff.h

File: src/openrct2/scripting/ScEntity.hpp [hunk 12]
```diff
@@ -1085,16 +1085,18 @@ namespace OpenRCT2::Scripting
...
-                    case STAFF_TYPE_MECHANIC:
+                    case StaffType::Mechanic:
                         return "mechanic";
-                    case STAFF_TYPE_SECURITY:
+                    case StaffType::Security:
                         return "security";
-                    case STAFF_TYPE_ENTERTAINER:
+                    case StaffType::Entertainer:
                         return "entertainer";
+                    case StaffType::Count:
```
Comment (duncanspumpkin) [inline src/openrct2/scripting/ScEntity.hpp:1098]: Add a default

**Comment by duncanspumpkin** [comment]
This needs a few changes still to make it align with the rest of the codebase.

File: src/openrct2-ui/windows/Staff.cpp (hunk 9 above)
Comment (tupaschoal) [reply src/openrct2-ui/windows/Staff.cpp:892]: I asked for th removal of default because it needs to use all enumerations otherwise it errors out, don't you think it looks clearer? This way we can see what actually is being handled or not

File: src/openrct2/peep/Peep.h (hunk 11 above)
Comment (tupaschoal) [reply src/openrct2/peep/Peep.h:63]: I guess he did it because he needed it for the member variable on peep

**Comment by ryan-bello** [comment]
Should I submit a new pr with these changes?

**Comment by tupaschoal** [comment]
Don't worry for now, let me and @duncanspumpkin converge on it first.

**Comment by ryan-bello** [comment]
Got it, sounds good

**Comment by Broxzier** [comment]
For the next PR, try creating a new branch instead of working on develop directly. It's good practice to keep develop the same as upstream, and branches allow you to work on different things and switch between them easily.

**Comment by ryan-bello** [comment]
Sure thing, will do!
